### PR TITLE
Add support for cracking Monero wallets

### PR DESCRIPTION
**Pull request by kholia**
NOTE: I still have to clean up this code!

This fixes https://github.com/magnumripper/JohnTheRipper/issues/3144.

### CODE REVIEW

**Comment by kholia** [comment]
Things to do,

* Reuse the existing Keccak code we have.

* Formatting, and whitespace cleanups.

**Comment by kholia** [comment]
This is a very slow hash,

```
$ ../run/john --test --format=monero  # i7-6600U CPU
Will run 4 OpenMP threads
Benchmarking: monero, monero Wallet [AES+ChaCha+Various 64/64]... (4xOMP) DONE
Raw:	9.1 c/s real, 2.4 c/s virtual
```

File: src/monero_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,270 @@
+/*
+ * Format for cracking Monero wallet hashes.
+ *
+ * This software is Copyright (c) 2017, Dhiru Kholia <dhiru [at] openwall.com>,
```
Comment (magnumripper) [inline src/monero_fmt_plug.c]: It's 2018!

File: src/monero_fmt_plug.c [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+#include "params.h"
+#include "options.h"
+#include "chacha.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "monero"
+#define FORMAT_NAME             "monero Wallet"
+#define FORMAT_TAG              "$monero$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "AES+ChaCha+Various 64/" ARCH_BITS_STR
```
Comment (magnumripper) [inline src/monero_fmt_plug.c]: We don't use to have `+` in algo names. I'd prefer spaces or perhaps `/`

File: src/monero_fmt_plug.c [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+{
+	saved_len[index] = strnzcpyn(saved_key[index], key, PLAINTEXT_LENGTH + 1);
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+void cn_slow_hash(const void *data, size_t length, char *hash);
```
Comment (magnumripper) [inline src/monero_fmt_plug.c]: Perhaps create (and source) a `slow_hash.h`?

File: src/monero_fmt_plug.c (hunk 2 above)
Comment (magnumripper) [reply src/monero_fmt_plug.c]: ...and perhaps it should say "pseudo AES" or something, since it's not a normal one.

**Comment by kholia** [comment]
I have addressed your code comments, thanks!

I am now testing cracking of Unicode passwords.

**Comment by kholia** [comment]
Update,

* I have added a Unicode test vector now.

* Our BLAKE2 implementation was somehow not working. Added BLAKE2 implementation from Monero project.

I don't like the duplicated implementations. Hopefully, we will be able to unify them somehow.

**Comment by kholia** [comment]
Monero wallet code base also supports `legacy` wallets. Support for them needs to be added and tested.

**Comment by kholia** [comment]
Maybe we don't need to support the old wallet format. Even the oldest downloadable release of Monero (i.e. 0.9.1 Hydrogen Helix) uses the new JSON wallet format.

```
$ head monero2john.py
...
# + Tested with monero-gui-v0.11.1.0 on Fedora 27.
# + Tested with monero.linux.x64.v0-9-0-0.tar.bz2 (from Jan, 2016) on Fedora 27.
```

**Comment by magnumripper** [comment]
> Mapping the new Keccak "API" to our existing Keccak implementation doesn't look trivial.

In what way was it not trivial? I just had a (very) quick look and I didn't see anything weird. Can't you just write wrapper macros?

> Our BLAKE2 implementation was somehow not working. Added BLAKE2 implementation from Monero project.

Do they produce different hashes? This should be investigated. I see there are flavours of Blake2... Here's test vectors for all of them: https://github.com/BLAKE2/BLAKE2/tree/master/testvectors

You have my "permission" to merge this as-is, although I'd rather see these issues fixed first (or it probably wont happen at all).

**Comment by kholia** [comment]
I have opened an issue about this code duplication stuff, and assigned it to myself.

**Pull request merged**